src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -48,6 +48,76 @@ typedef struct uct_cuda_ipc_remote_cache {
 
 uct_cuda_ipc_remote_cache_t uct_cuda_ipc_remote_cache;
 
+#if HAVE_CUDA_FABRIC
+static UCS_F_ALWAYS_INLINE int
+uct_cuda_ipc_rem_mpool_cache_hash_equal(CUmemFabricHandle *key1,
+                                        CUmemFabricHandle *key2)
+{
+    return !memcmp(key1, key2, sizeof(*key1));
+}
+
+static UCS_F_ALWAYS_INLINE khint64_t
+uct_cuda_ipc_rem_mpool_cache_hash_func(CUmemFabricHandle *key)
+{
+    return kh_int64_hash_func((khint64_t)key->data);
+}
+
+KHASH_INIT(cuda_ipc_rem_mpool_cache, CUmemFabricHandle*,
+           CUmemoryPool*, 1, uct_cuda_ipc_rem_mpool_cache_hash_func,
+           uct_cuda_ipc_rem_mpool_cache_hash_equal);
+
+typedef struct uct_cuda_ipc_rem_mpool_cache {
+    khash_t(cuda_ipc_rem_mpool_cache) hash;
+    ucs_recursive_spinlock_t          lock;
+} uct_cuda_ipc_rem_mpool_cache_t;
+
+uct_cuda_ipc_rem_mpool_cache_t uct_cuda_ipc_rem_mpool_cache;
+
+static ucs_status_t
+uct_cuda_ipc_get_rem_mpool_cache(CUmemFabricHandle *fabric_handle,
```

Comment:
can we reuse the existing IPC remote cache instead of adding a new one?

Reply:
IPC mechanism with memory pools involves two steps in the import phase. One expensive step to import the actual memory pool to allows local GPU to be capable of accessing locally mapped memory pool; and another low overhead step to import remote pointer. The exporting process is expected to make several allocations and deallocations from the memory pool and pack associated pointer and memory pool handle in the rkey. Importing process checks for remote memory pool handle and performs a mapping only when not present. 

The existing IPC remote cache relies on handle as key and returns a VA as value and directly uses offsets from the VA range for copy purposes. This doesn't translate to remote memory pool cache as we need imported memory and use that with the exported pointer [here](https://github.com/openucx/ucx/pull/9982/files#diff-561a7e4fa208245423bb161bc7a1cc05fbce0d65a10ef13c77e366f38c541826R256) to import associated VA. Hence there isn't a one-to-one mapping between the two remote caches.

Reply:
maybe we could still unite them, if the hash key will contain the handle type?

Reply:
@yosefe Today, we use exporter-side base pointer to lookup cache and perform import operation if base poiner isn't found. This works fine for legacy cuda-ipc because there is one independent backing for a given allocation. Two allocations can never have a common backing. So performing expensive import operation for each new unique base pointer is justified (we also need to check that the handle for given base pointer matches with cached handle to detect VA recycling). 

For mempools, many sub-allocations will have the same common backing (the memory pool). The first time we encounter a base pointer corresponding to a sub-allocation from the mempool, we import the whole mempool and then use that imported mempool and exported data pointer to get local translation of the associated base pointer of the sub-allocation at the importer side. If another sub-allocation from the same mempool is encountered at the importing process, if we use existing IPC remote cache logic, we will re-import an already imported mempool (paying unnecessary cost) and get local translation. Instead, it is better to check if a given mempool has been imported. Therefore we have a new cache that takes exported handle as input and returns imported handle as output. This way, any new sub-allocation from already imported memory pool will avoid import operation and just data pointer is imported. Since the old cache takes base pointer as input and the new cache introduced takes exporter handle as input, I don't see an immediate way of unifying the cache as we do need a common key type for a common cache.